misc-tools/dj_utils.py

```diff
@@ -91,19 +104,64 @@ def upload_file(name: str, apifilename: str, file: str, data: dict = {}):
         RuntimeError when the HTTP status code is non 2xx.
     '''
 
-    global ca_check
-    files = [(apifilename, open(file, 'rb'))]
+    if os.path.isdir(domjudge_webapp_folder_or_api_url):
+        return api_via_cli(name, 'POST', data, {apifilename: file})
+    else:
+        global ca_check
+        files = [(apifilename, open(file, 'rb'))]
+
+        url = f'{domjudge_webapp_folder_or_api_url}/{name}'
+
+        try:
+            response = requests.post(
+                url, files=files, headers=headers, data=data, verify=ca_check)
+        except requests.exceptions.SSLError as e:
+            ca_check = not confirm(
+                "Can not verify certificate, ignore certificate check?", False)
+            if ca_check:
+                print('Can not verify certificate chain for DOMserver.')
+                exit(1)
+            else:
+                response = requests.post(
+                    url, files=files, headers=headers, data=data, verify=ca_check)
 
-    url = f'{api_url}/{name}'
+    return parse_api_response(name, response)
 
-    try:
-        response = requests.post(url, files=files, headers=headers, data=data, verify=ca_check)
-    except requests.exceptions.SSLError as e:
-        ca_check = not confirm("Can not verify certificate, ignore certificate check?", False)
-        if ca_check:
-            print('Can not verify certificate chain for DOMserver.')
-            exit(1)
-        else:
-            response = requests.post(url, files=files, headers=headers, data=data, verify=ca_check)
 
-    return parse_api_response(name, response)
+def api_via_cli(name: str, method: str = 'GET', data: dict = {}, files: dict = {}):
+    '''Perform the given API request using the CLI
+
+    Parameters:
+        name (str): the endpoint to call
+        method (str): the method to use. Either GET or POST
+        data (dict): the POST data to use. Only used when method is POST
+        files (dict): the files to use. Only used when method is POST
+
+    Returns:
+        The parsed endpoint contents.
+
+    Raises:
+        RuntimeError when the command exit code is not 0.
+    '''
+
+    command = [
+        f'{domjudge_webapp_folder_or_api_url}/bin/console',
+        'api:call',
+        '-m',
+        shlex.quote(method),
+    ] + list(itertools.chain(*[
```

Comment:
Instead of this I would just write something simpler like:

```
command = [
        f'{domjudge_webapp_folder_or_api_url}/bin/console',
...
        shlex.quote(method),
]
for item in data:
    command.extend(['-d', f'{shlex.quote(item)}={shlex.quote(data[item])}']
...
command.append(shlex.quote(name))
```

Reply:
So smart